File: in_Java/0133 Clone Graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    Map<Integer, Node> hm = new HashMap<>();
    public Node cloneGraph(Node node) {
        return dfs(node);
    }
    private Node dfs(Node n){
        if(n == null) return n;
        if(hm.containsKey(n.val)) return hm.get(n.val);
        Node nn = new Node(n.val, new ArrayList<Node>());
        hm.put(n.val, nn);
        for(Node nei : n.neighbors){
            nn.neighbors.add(dfs(nei));
        }
        return nn;
    }
}

File: in_Python/0133 Clone Graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        if not node:
            return node
        d = {node : Node(node.val)}
        q = collections.deque([node])
        while q:
            r = q.popleft()
            for nex in r.neighbors:
                if nex not in d:
                    d[nex] = Node(nex.val)
                    q.append(nex)
                d[r].neighbors.append(d[nex])
        return d[node]